test/jdk/java/net/httpclient/CancelledResponse2.java

```diff
@@ -0,0 +1,258 @@
+        public HttpResponse.BodySubscriber<String> apply(HttpResponse.ResponseInfo rinfo) {
+            assert !cancelled.get();
+            return new CancellingBodySubscriber(expected, cancelled);
+        }
+    }
+
+
+    static class CancellingBodySubscriber implements HttpResponse.BodySubscriber<String> {
+        private final String expected;
+        private final CompletableFuture<String> result;
+        private Flow.Subscription subscription;
+        final AtomicInteger index = new AtomicInteger();
+        final AtomicBoolean cancelled;
+        CancellingBodySubscriber(String expected, AtomicBoolean cancelled) {
+            this.cancelled = cancelled;
+            this.expected = expected;
+            result = new CompletableFuture<>();
+        }
+
+        @Override
+        public CompletionStage<String> getBody() {
+            return result;
+        }
+
+        @Override
+        public void onSubscribe(Flow.Subscription subscription) {
+            this.subscription = subscription;
+            subscription.request(1);
+        }
+
+        @Override
+        public void onNext(List<ByteBuffer> item) {
+            //if (result.isDone())
+            // Max Delay is 180ms as there is 160 characters in response which gives at least 160ms in some test cases and
+            // allows a response to complete fully with a lee-way of 20ms in other cases. Otherwise, response body is
+            // usually partial. Each character is written by the server handler with a 1ms delay.
+            delay();
+            for (ByteBuffer b : item) {
+                while (b.hasRemaining() && !result.isDone()) {
+                    int i = index.getAndIncrement();
+                    char at = expected.charAt(i);
+                    byte[] data = new byte[b.remaining()];
+                    b.get(data); // we know that the server writes 1 char
+                    String s = new String(data);
+                    char c = s.charAt(0);
+                    System.err.print(c);
+                    if (c != at) {
+                        Throwable x = new IllegalStateException("char at "
+                                + i + " is '" + c + "' expected '"
+                                + at + "' for \"" + expected +"\"");
+                        out.println("unexpected char received, cancelling");
+                        subscription.cancel();
+                        result.completeExceptionally(x);
+                        return;
+                    }
+                }
+                System.err.println();
+            }
+            if (index.get() > 0 && !result.isDone()) {
+                // we should complete the result here, but let's
+                // see if we get something back...
+                out.println("Cancelling subscription after reading " + index.get());
+                cancelled.set(true);
+                subscription.cancel();
+                result.completeExceptionally(new CancelException());
```

Comment:
I believe that's the exception we expect to find in the `HttpResponse`. IIRC the original test was checking for that.

Reply:
Right, but calling `subscription.cancel()` causes `Stream.cancelImpl()` to be called. That in turn causes the `HttpResponse` (or the variable result in your snippet above) to complete exceptionally with an IOException which has the message "Stream x cancelled". I think that means that the call to `completeExceptionally(new CancelException())` has no effect because `subscription.cancel()` triggers a call to `cancelImpl()`